glass/src/lib/native/cpp/support/ExpressionParser.cpp

```diff
@@ -0,0 +1,374 @@
+    return wpi::unexpected("No value");
+  }
+
+  // Intentionally leaves the result value on top of valStack so unmatched
+  // closing parentheses work
+  return valStack.top();
+}
+
+template <typename V>
+wpi::expected<V, std::string> ParseExpr(Lexer& lexer, bool insideParen) {
+  std::stack<Operator> operStack;
+  std::stack<V> valStack;
+
+  bool prevWasOp = true;
+  TokenType prevType = TokenType::Add;
+
+  while (true) {
+    Token token = lexer.NextToken();
+
+    bool wasOp = false;
+    switch (token.type) {
+      case TokenType::End:
+        goto end;
+      case TokenType::Number: {
+        // Check for two numbers in a row (ex: "2 4"). Implicit multiplication
+        // is probably not what the user intended in this case, so give them an
+        // error.
+        if (prevType == TokenType::Number) {
+          return wpi::unexpected("Missing operator");
+        }
+
+        // Implicit multiplication. Ex: "2(4 + 5)"
+        if (!prevWasOp) {
+          operStack.push(Operator::Multiply);
+        }
+
+        auto value =
+            ValueFromString<V>(std::string_view(token.str, token.strLen));
+        if (value) {
+          valStack.push(value.value());
+        } else {
+          return wpi::unexpected("Invalid number");
+        }
+
+        break;
+      }
+
+      case TokenType::OpenParen: {
+        // Implicit multiplication
+        if (!prevWasOp) {
+          operStack.push(Operator::Multiply);
+        }
+
+        wpi::expected<V, std::string> result = ParseExpr<V>(lexer, true);
+        if (!result) {
+          return result;
+        }
+        valStack.push(result.value());
+
+        TokenType nextType = lexer.NextToken().type;
+        if (nextType != TokenType::CloseParen) {
+          if (nextType == TokenType::End) {
+            goto end;  // Act as if closed at end of expression
+          }
+          return wpi::unexpected("Expected )");
+        }
+        break;
+      }
+
+      case TokenType::CloseParen: {
+        if (insideParen) {
+          lexer.Repeat();
+          goto end;
+        }
+
+        // Acts as if there was open paren at start of expression. EvalAll will
+        // clear both stacks, and leave the result value on top of valStack.
+        // This makes sure everything inside the parentheses is evaluated first
+        wpi::expected<V, std::string> result = EvalAll<V>(operStack, valStack);
+        if (!result) {
+          return result;
+        }
+        break;
+      }
+
+      case TokenType::Error:
+        return wpi::unexpected(std::string("Unexpected character: ")
+                                   .append(token.str, token.strLen));
+
+      default:
+        Operator op = GetOperator(token.type);
+        if (op == Operator::None) {
+          lexer.Repeat();
+          goto end;
+        }
+        if (op == Operator::Subtract && prevWasOp) {
+          op = Operator::Negate;
+          // Dummy left-hand side for negation
+          valStack.push(0.0);
```

Comment:
Have we checked that this `0.0` literal works fine for the `int64_t` case? It might be okay, but I'd like to make sure.

Reply:
Yes it does work :+1: